### ISSUE TITLE
Offline + Changing Device Time Problem

### ISSUE THREAD

**Issue by sunnykinger** [AllanHasegawa/Tempo#9]
I tried a few things on the sample application which displays the System time and Tempo time.
When you run the app (Internet Connected), it shows the time correctly for Tempo time even if you go to Settings of Phone and change device time. Now if you turn off the Internet (Mobile data and Wifi) and run the app again, still it will work fine. Now closing the app again with the Internet off and then going to settings and changing device time, now try and open the app, Tempo time show blank because some exception is thrown. Is this the normal case or is it really a bug?

**Comment by AllanHasegawa** [comment]
This is a bug. Do you have a stack trace for the exception?

**Comment by sunnykinger** [comment]
`fun isCacheValid(cache: TimeSourceCache): Boolean {
            val estimatedBootTime = deviceClocks.estimatedBootTime()
            val cacheEstimatedBootTime = cache.estimatedBootTime
            return Math.abs(cacheEstimatedBootTime - estimatedBootTime) <= 5000L
        }`

This code in the library returns false because when we change the device time, this condition is not fulfilled.

Exception
`io.tempo.internal.AndroidSntpClient$InetException: Error getting the host (time.google.com) inet address.
        at io.tempo.internal.AndroidSntpClient.queryHostAddress(AndroidSntpClient.kt:54)
        at io.tempo.time_sources.SlackSntpTimeSource$requestTime$1.subscribe(SlackSntpTimeSource.kt:56)
        at io.reactivex.internal.operators.single.SingleCreate.subscribeActual(SingleCreate.java:39)
        at io.reactivex.Single.subscribe(Single.java:3603)
        at io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
        at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
        at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
        at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)`

**Comment by AllanHasegawa** [comment]
Thank you for digging into the issue.

The problem is that the only clock that can be trusted on an Android Device is the `SystemClock.elapsedRealTime()` [link](https://developer.android.com/reference/android/os/SystemClock#elapsedRealtime()). The downside is that clock resets on every boot. This is the main issue here, as, when Tempo is initialized, it needs to now if the cached time measure and this clock are in sync, or in other words, has this clock reseted between the last trusted time measured and the app start?

Currently Tempo uses `System.currentTimeMillis()` for that check by recording the time of each boot. And by using it, the user can change the device time and change Tempo's behaviour, which is bad.

So basically we need a way to uniquely identify (`id`) each boot, and tie that `id` with each trusted time measurement. The problem is how 🤔 

I've played with `android.permission.RECEIVE_BOOT_COMPLETED` but I'm not yet convinced this is the correct approach. I remember having issues with some devices with an earlier implementation.

So, current status: Still investigating on how to solve it properly.

**Comment by AllanHasegawa** [comment]
@sunnykinger Thanks again for the detailed issue report 🙏 

The PR #10 fixes this issue for Android devices with API Level 24+. It uses the Settings Global [BOOT_COUNT](https://developer.android.com/reference/android/provider/Settings.Global#BOOT_COUNT) to uniquely identify each boot. That setting should be safe to use in a non-tampered device.

Unfortunately I couldn't find a reliable way to solve it for older devices, therefore, the old behaviour persists for those devices as a fallback mechanism.

**Comment by sunnykinger** [comment]
Great @AllanHasegawa
